pkg/spanconfig/spanconfigstore/systemspanconfigstore.go

```diff
@@ -0,0 +1,184 @@
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package spanconfigstore
+
+import (
+	"sort"
+
+	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/spanconfig"
+	"github.com/cockroachdb/errors"
+)
+
+// systemSpanConfigStore is an in-memory data structure to store system span
+// configurations. It is not safe for concurrent use.
+type systemSpanConfigStore struct {
+	store map[spanconfig.SystemTarget]roachpb.SpanConfig
+}
+
+// newSystemSpanConfigStore constructs and returns a new systemSpanConfigStore.
+func newSystemSpanConfigStore() *systemSpanConfigStore {
+	store := systemSpanConfigStore{}
+	store.store = make(map[spanconfig.SystemTarget]roachpb.SpanConfig)
+	return &store
+}
+
+// apply takes an incremental set of system span config updates and applies them
+// to the underlying store. It returns a list of targets deleted/records added.
+//
+// TODO(arul): Even though this is called from spanconfig.Store.Apply, which
+// accepts a dryrun field, we don't accept a a dry run option here because it's
+// unused; instead, we plan on removing it entirely.
+func (s *systemSpanConfigStore) apply(
+	updates ...spanconfig.Update,
+) (deleted []spanconfig.SystemTarget, added []spanconfig.Record, _ error) {
+	if err := s.validateApplyArgs(updates); err != nil {
+		return nil, nil, err
+	}
+
+	for _, update := range updates {
+		if update.Addition() {
+			conf, found := s.store[update.Target.GetSystemTarget()]
+			if found {
+				if conf.Equal(update.Config) {
+					// no-op.
+					continue
+				}
+				deleted = append(deleted, update.Target.GetSystemTarget())
+			}
+
+			s.store[update.Target.GetSystemTarget()] = update.Config
+			added = append(added, spanconfig.Record(update))
+		}
+
+		if update.Deletion() {
+			_, found := s.store[update.Target.GetSystemTarget()]
+			if !found {
+				continue // no-op
+			}
+			delete(s.store, update.Target.GetSystemTarget())
+			deleted = append(deleted, update.Target.GetSystemTarget())
+		}
+	}
+
+	return deleted, added, nil
+}
+
+// combine takes a key and an associated span configuration and combines it with
+// all system span  configs that apply to the given key.
+func (s *systemSpanConfigStore) combine(
+	key roachpb.RKey, config roachpb.SpanConfig,
+) (roachpb.SpanConfig, error) {
+	_, tenID, err := keys.DecodeTenantPrefix(roachpb.Key(key))
+	if err != nil {
+		return roachpb.SpanConfig{}, err
+	}
+
+	hostSetOnTenant, err := spanconfig.MakeTenantKeyspaceTarget(
+		roachpb.SystemTenantID, tenID,
+	)
+	if err != nil {
+		return roachpb.SpanConfig{}, err
+	}
+	// Construct a list of system targets that apply to the key given its tenant
+	// prefix.
```

Comment:
nit: Maybe explicity write out the three targets we want to check against, as part of this comment? Might be helpful for future readers even though the constructor methods are quite intuitively named now.